<a href="https://colab.research.google.com/github/Wipersee/optimization_dfp/blob/main/Kursach.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<left>
    <img src="https://kpi.ua/files/images/kpi.png" width="300" alt="kpi logo"  />
</left>

## Метод Девідона-Флетчера-Паула
### Виконав студент групи КМ-81 Піткевич Ілля
**Мінімізувати функцію Розенброка методом ДФП**

#### Зміст
*   Постановка задачі
*   Теоретична частина
*   Розрахункова частина





## Постановка задачі:
Дослідити збіжність метода ДФП при мінімізації функції Розенброка в залежності від:


1.   Величини кроку h при обчисленні похідних
2.   Схеми обчислення похідних
3.   Виду методу одновимірного пошуку (ДСК-Пауелла або Золотого перетину)
4.   Точності методу одновимірного пошуку
5.   Значення параметру в алгоритмі Свена
6.   Вигляду критерія закінчення
7.   Наявності рестартів



## Теоретична частина
**Метод Девідона-Флетчера-Пауелла**

Метод ДФП заснований на використанні ідей методу Ньютона і методів, які використовують зв'язані напрямки. Матриця напрямків А обчислюється таким чином, щоб для квадратичної цільової функції в межі після n кроків вона дорівнювала $H^{-1}$. Початкова матриця А береться рівній одиничній матриці І. 

**Алгоритм методу**

Вихідні дані: точка $x^{0}$, параметри закінчення $\varepsilon_1, \varepsilon_2$.


1.   Задати координати початкової точки $x^{k}$. Обчислити значення $\Delta f(x^{k})$. $A^{0}=I$
2.   Обчислити $\lambda^{k}$ за допомогою одновимірного методу пошуку
3.    Обчислити точку $x^{k+1}=x^{k} - \lambda^{k}*A*\Delta f(x^{k})$
4.    Обчислити $f(x^{k}), \Delta f(x^{k})$
5.    Перевірити виконання критерію закінчення. Якщо виконується - кінець пошуку, якщо ні шукаємо $A$
6.    $A^{k} = A^{k-1}+ \frac{\Delta x^{k-1} \Delta {x^{(k-1)}}^T}{\Delta {x^{(k-1)}}^{T} \Delta g^{k-1}} - \frac{A^{k-1}\Delta g^{k-1}\Delta {g^{(k-1)}}^{T} A^{k-1} }{\Delta {g^{(k-1)}}^{T} A^{k-1} \Delta g^{k-1} }$



## Розрахункова частина

### Безумовна оптимізація

1. Оцінка залежності збіжності функції від кроку обчислення похідних

* Схема обчислення похідних – центральна
* Метод одновимірного пошуку – ДСК-Пауелла
* Точність методу одновимірного пошуку – 0.01
* Параметр у алгоритмі Свенна – 0.1
* Точність методу – 0.001

Значення кроку обчислення похідних  | Викликів функції | Знайдена точка
------------- | ------------- | ------------
0.001  | 287 | [0.99934026 0.99867854]
0.0001  | 308 | [0.99999082 0.99998254]
0.00001  | 308 | [0.99999751 0.99999591]
0.000001  | 308 | [0.99999757 0.99999604]

**Найбільш оптимальним значенням кроку обчислення похідних є 0.00001.**
2. Оцінка залежності збіжності функції від схеми обчислення похідних
* Крок обчислення похідних – 0.00001
* Метод одновимірного пошуку – ДСК-Пауелла
* Точність методу одновимірного пошуку – 0.01
* Параметр у алгоритмі Свенна – 0.1
* Точність методу – 0.001

Схема обчислення похідних  | Викликів функції | Знайдена точка
------------- | ------------- | ------------
Центральна  | 308 | [0.99999751 0.99999591]
Ліва  | 1149 | [1.00302603 1.00608199]
Права  | 1024 | [0.9971594  0.99431986]

3. Оцінка залежності збіжності функції від методу одновимірного пошуку

* Крок обчислення похідних – 0.00001
* Схема обчислення похідних – центральна
* Точність методу одновимірного пошуку – 0.01
* Параметр у алгоритмі Свенна – 0.1
* Точність методу – 0.001

Метод одновимірного пошуку  | Викликів функції | Знайдена точка
------------- | ------------- | ------------
ДСК Пауелла  | 308 | [0.99999751 0.99999591]
Золотий перетин  | 537 | [0.99998812 0.99997639]

**Вибираємо ДСК-Паулла**

4. Оцінка залежності збіжності функції від точності методу одновимірного пошуку

* Крок обчислення похідних – 0.00001
* Схема обчислення похідних – центральна
* Метод одновимірного пошуку – ДСК-Пауелла
* Параметр у алгоритмі Свенна – 0.1
* Точність методу – 0.001

Точність одновимірного пошуку  | Викликів функції | Знайдена точка
------------- | ------------- | ------------
0.01  | 308 | [0.99999751 0.99999591]
0.001  | 292 | [1.00000801 1.00001541]
0.0001  | 321 | [0.99998665 0.99997419]

**Як бачимо, доцільно використовувати 0.001.**

5. Оцінка залежності збіжності функції від значення параметру в алгоритмі Свенна

* Крок обчислення похідних – 0.00001
* Схема обчислення похідних – центральна
* Метод одновимірного пошуку - ДСК-Пауелла 
* Точність методу одновимірного пошуку – 0.001
* Точність методу – 0.001

Значення параметру в алгоритмі Свена  | Викликів функції | Знайдена точка
------------- | ------------- | ------------
0.01  | 292 | [1.00000801 1.00001541]
0.001  | 543 | [0.99999874 0.99999762]
0.0001  | 538 | [1.00002352 1.00004496]

**Будемо використовувати 0.01**

6. Оцінка залежності збіжності функції від точності методу для безумовної оптимізації

* Крок обчислення похідних – 0.00001
* Схема обчислення похідних – центральна
* Метод одновимірного пошуку - ДСК-Пауелла 
* Точність методу одновимірного пошуку – 0.001
* Параметр у алгоритмі Свенна – 0.01

Точність методу  | Викликів функції | Знайдена точка
------------- | ------------- | ------------
0.1  | 93 | [0.9407137  0.88480521]
0.01  | 292 | [1.00000801 1.00001541]
0.001  | 292 | [1.00000801 1.00001541]
0.0001  | 323 | [1.00000007 1.00000015]
0.00001  | 323 | [1.00000007 1.00000015]

**Виберемо точність 0.001**


### Умовна оптимізація

Задамо допустиму область:
$-x_1^{2}-x_2^{2}+2.3 \ge 0$

Тоді штрафна функція набуде вигляду:

$ (1-x)^{2} + 100*(-x^{2}+y)^{2} + R(-x^{2}-y^{2}+2.3)$

**Точка лежить поза областю**

R | $X_1$ | $X_2$ | Кіл-сть викликів
--- | --- | ---- | ----
0.1 | 1.03842578 | 1.07845451 | 615 |
1 | 1.0462301 | 1.09474909 | 511 |
2 | 1.04674307 | 1.09582439 | 783 |
5 | 1.04705548 | 1.09647951 | 421 |

**Точка лежить в області**

R | $X_1$ | $X_2$ | Кіл-сть викликів
--- | --- | ---- | ----
0.1 |1.03842578 | 1.07845452 | 485 |
1 | 1.0462301 | 1.09474909 | 388 |
2 | 1.04674307 | 1.09582439 | 369 |
10 | 1.0471604 | 1.09669958 | 427 |


Зробимо область невипуклою, покладанням додаткового обмеження:
$x_1^{2}+x_2^{2} - 0.7 \ge 0$

Тоді штрафна функція набуде вигляду:
$ (1-x)^{2} + 100*(-x^{2}+y)^{2} + R(-x^{2}-y^{2}+2.3) + R(x_1^{2}+x_2^{2} - 1.1)^{2}$

**Точка лежить поза областю**

R | $X_1$ | $X_2$ | Кіл-сть викликів
--- | --- | ---- | ----
0.1 | 0.9547169 | 0.91133122 | 545 |
1 | 0.94778695 | 0.89812316 | 481 |
10 | 0.94695246 | 0.89653915 | 480 |


**Точка лежить в області**

R | $X_1$ | $X_2$ | Кіл-сть викликів
--- | --- | ---- | ----
0.1 | 0.95471689 | 0.91133121 | 420 |
1 | 0.94778694 | 0.89812316 | 234 |
10 | 0.94695246 | 0.89653915 | 284 |

In [ ]:
!pip install numpy
!pip install scipy

In [ ]:
import math 
import numpy as np
from scipy.optimize import approx_fprime
from pprint import pprint

In [ ]:
calls_of_functions = 0
derivative_calculation_step = 0.00001
left_gold_section=0.382
right_gold_section=0.618

Функція Розенброка
$f(x_1,x_2) = 100(x_1^{2} - x_2)^{2} + (x_1 - 1)^{2}$, яка має мінімум в точці $x_0 = (1, 1)$

In [ ]:
def fun(x):
  global calls_of_functions
  calls_of_functions += 1
  return 100*(x[0]**2 - x[1])**2 + (x[0]-1)**2

Пишемо функції для  для показу результатів, суми, різниці та віднімання точок.

In [ ]:
def add(x, y):
    res = []
    for i in range(len(x)):
        res.append(x[i] + y[i])
    return res


def sub(x, y):
    res = []
    for i in range(len(x)):
        res.append(x[i] - y[i])
    return res


def mults(x, n):
    res = []
    for i in range(len(x)):
        res.append(x[i] * n)
    return res


def print_summary(calls_of_functions, iteration, restart, A, x):
    print("Кінець програми")
    print(f"Функція була обчислена: {calls_of_functions} разів")
    print(f"ДФП ітерацій проведено: {iteration}")
    print(f"Рестартів у ДФП проведено: {restart}")
    print(f"Обернена матриця Гессе:")
    pprint(hesse(x))
    print(f"Матриця напрямків А:")
    pprint(A)


def print_interim_results(x, calls_of_functions, iteration, restart):
    print(f"Нова точка: {x}")
    print(f"Функція була обчислена: {calls_of_functions} разів")
    print(f"Ітерація ДФП: {iteration}")


Реалізуємо функції для похідних центрального, правого та лівого типів 

Джерело: http://aco.ifmo.ru/el_books/numerical_methods/lectures/glava1.html

*МЕТОДИ ОПТИМІЗАЦІЇ.РЕКОМЕНДАЦІЇ ДО ВИКОНАННЯ КУРСОВОЇ РОБОТИ* 15-18 стор.

In [ ]:
def derivative(x, n):  # Центр тип
    h = []
    for i in range(len(x)):
        if i == n:
            h.append(derivative_calculation_step)  # Точність пошуку похідної
        else:
            h.append(0) # Так як часткова похідна, то нам потрібно обнулиити другу змінну
    return (fun([x[0] + h[0], x[1] + h[1]]) - fun([x[0] - h[0], x[1] - h[1]])) / (
        2 * h[n]
    )


def derivative2(x, n):  # Правий тип
    h = []
    for i in range(len(x)):
        if i == n:
            h.append(derivative_calculation_step)  # Точність пошуку похідної
        else:
            h.append(0)
    return (fun([x[0] + h[0], x[1] + h[1]]) - fun([x[0], x[1]])) / (h[n])


def derivative3(x, n):  # Лівий тип
    h = []
    for i in range(len(x)):
        if i == n:
            h.append(derivative_calculation_step)  # Точність пошуку похідної
        else:
            h.append(0)
    return (fun([x[0], x[1]]) - fun([x[0] - h[0], x[1] - h[1]])) / (h[n])

Реалізуємо функції для знаходження градієнту, норми та дельт в ДФП

Функція для підрахунку Гесіана https://gist.github.com/jgomezdans/3144636

Дока по approx_fprime: https://scipy.github.io/devdocs/reference/generated/scipy.optimize.approx_fprime.html

In [ ]:
def gradient(x):
    grad = []
    for i in range(len(x)):
        grad.append(derivative(x, i))
    return grad


def norm(s1):
    normas = 0
    for i in range(len(s1)):
        normas += s1[i] ** 2
    return math.sqrt(normas)


def hesse ( x0, epsilon=1.e-5, linear_approx=False, *args ):
    f1 = approx_fprime( x0, fun, epsilon, *args) 
    if linear_approx:
        f1 = np.matrix(f1)
        return f1.transpose() * f1    
    n = x0.shape[0]
    hessian = np.zeros ( ( n, n ) )
    xx = x0
    for j in range( n ):
        xx0 = xx[j] # Заповнюємо старими значеннями
        xx[j] = xx0 + epsilon
        f2 = approx_fprime( x0, fun, epsilon, *args ) 
        hessian[:, j] = (f2 - f1)/epsilon
        xx[j] = xx0 # Restore initial value of x0        
    return np.linalg.inv(hessian)

def find_deltas(A, x0, x1):
    deltag = np.array(sub(gradient(x1), gradient(x0)))[np.newaxis] # Різниця g
    deltax = sub(x1, x0) # Різниця x

    deltax = np.array(deltax)[np.newaxis] # Розширюємо deltax
    deltaxT = np.array(deltax).T # Транспонуємо detlax
    deltagT = np.array(deltag).T# Транспонуємо detltag

    deltax, deltaxT = deltaxT, deltax
    deltag, deltagT = deltagT, deltag

    # Шукаємо перший дріб в формулі
    first = np.dot(deltax, deltaxT) # Перемножуємо
    second = np.dot(deltaxT, deltag)
    resOne = first / second

    # Шукаємо другий дріб
    first = np.dot(A, deltag)
    second = np.dot(first, deltagT)
    third = np.dot(second, A)
    temp = third

    first = np.dot(deltagT, A)
    second = np.dot(first, deltag)

    resTwo = temp / second
    return resOne, resTwo # Вертаємо кортеж з результатами

Пишемо функцію Свена та допоміжну функцію(для підрахунку точки х(k+1)

$x(k+1) = x_k \pm \Delta*2^{k}$ - формула для знаходження інтервалу невизначенності, при кроці lmb в функції Svenn

In [ ]:
def calculate_x(x0, A, grad, lmb):
    lmbA = np.dot(lmb, A)
    lmbAgrad = np.dot(lmbA, grad)
    calc = x0 - lmbAgrad
    return calc


def svenn(x0, grad, lmb, delta, A):
    f0 = fun(calculate_x(x0, A, grad, lmb))
    if f0 < fun(
        calculate_x(x0, A, grad, lmb + delta)
    ):  # визначаємо куди рухаємось ліворуч чи праворуч
        delta = -delta
    x1 = lmb + delta  # власне рухаємось
    f1 = fun(calculate_x(x0, A, grad, x1))
    while f1 < f0:  # рухаємось доти, доки не почнемо рости
        delta *= 2
        lmb = x1
        x1 = lmb + delta
        f0 = f1
        f1 = fun(calculate_x(x0, A, grad, x1))
    a = lmb + delta / 2  # Знаходимо наш інтервал невизначенності
    b = lmb - delta / 2
    f0 = fun(calculate_x(x0, A, grad, lmb))
    f1 = fun(calculate_x(x0, A, grad, b))

    if f0 < f1:
        if a < b:
            return [a, b]
        else:
            return [b, a]
    elif f1 < f0:
        if lmb < x1:
            return [lmb, x1]
        else:
            return [x1, lmb]
    else:
        if lmb < b:
            return [lmb, b]
        else:
            return [b, lmb]

Реалізуємо метод одновимірного пошуку - ДСК-Паула.
Алгоритм:
1. Обчислюємо мінімум апроксимуючого поліному:
$ x_a = x_k + \frac{\Delta x (f(x_1)-f(x_3))}{2(f(x_1) - 2f(x_2) +f(x_3))}$
2. Крутимося доти, доки не виконається одне з умов закінчення пошуку:

  $|f(x_2) - f(x_m)| \le \varepsilon$

  $ |x_2-x_m| \le \varepsilon$

In [ ]:
def dscPowell(x0, grad, eps, lmb, delta, A):
    svenn_res = svenn(x0, grad, lmb, delta, A)  # Отримаємо інтервал, де лямбда
    a = svenn_res[0]
    b = svenn_res[1]
    xmin = (a + b) / 2
    f1 = fun(calculate_x(x0, A, grad, a))
    f2 = fun(calculate_x(x0, A, grad, xmin))
    f3 = fun(calculate_x(x0, A, grad, b))
    xApprox = xmin + ((b - xmin) * (f1 - f3)) / (
        2 * (f1 - 2 * f2 + f3)
    )  # Отримуємо апросимуючий поліном

    while (
        abs(xmin - xApprox) >= eps
        or abs(
            fun(calculate_x(x0, A, grad, xmin)) - fun(calculate_x(x0, A, grad, xApprox))
        )
        >= eps
    ):
        if xApprox < xmin:
            b = xmin
        else:
            a = xmin
        xmin = xApprox
        funcRes = [
            fun(calculate_x(x0, A, grad, a)),
            fun(calculate_x(x0, A, grad, xmin)),
            fun(calculate_x(x0, A, grad, b)),
        ]
        a1 = (funcRes[1] - funcRes[0]) / (xmin - a)
        a2 = ((funcRes[2] - funcRes[0]) / (b - a) - a1) / (b - xmin)
        xApprox = (a + xmin) / 2 - a1 / (2 * a2)
    return xmin


Пишемо функцію золотого перерізу

$x_1=a+0.382L$ &nbsp; 
$x_2=a+0.618L$

Повертає лямбду.

In [ ]:
def gold(a, b, eps, x0, grad, A):
    l = b - a
    x1 = a + l * left_gold_section
    x2 = a + l * right_gold_section
    while l > eps:
        if fun(calculate_x(x0, A, grad, x1)) < fun(calculate_x(x0, A, grad, x2)):
            b = x2
            x2 = x1
            l = b - a
            x1 = a + l * left_gold_section
        else:
            a = x1
            x1 = x2
            l = b - a
            x2 = a + l * right_gold_section
    print(f"Золотий переріз інтервал: [{round(a,4)},{round(b,4)}]")
    return [a, b]


def golden_lambda(x0, grad, eps, lmb, A):
    line = svenn(x0, grad, lmb, 0.1, A)
    line = gold(line[0], line[1], eps, x0, grad, A)
    lmb = (line[0] + line[1]) / 2
    return lmb

Пишемо реалізацию ДФП

In [ ]:
def dfp(x0, eps1, eps2):
    restart = 0
    iteration = 1
    lmb = 0.01
    A = np.eye(len(x0))  # Створюємо одиничну матрицу I

    print(x0)
    print("------------------------")
    while 1:
        grad = gradient(x0)  # шукаємо градієнт в точці

        if iteration > 0:
            if norm(grad) < eps1:  # Умова закінчення пошуку
                print_summary(calls_of_functions, iteration, restart, A, x1)
                break

        lmb = dscPowell(x0, grad, eps2, lmb, 0.01, A)  # Шукаємо лямбду за допомогою ДСК
        # lmb = golden_lambda(x0, grad, eps2, lmb, A)
        if lmb < 0:
            print("Рестарт")
            A = np.eye(len(x0))
            restart += 1

        x1 = calculate_x(x0, A, grad, lmb)  # Обчислюємо точку x1 за формулою 3

        resOne, resTwo = find_deltas(A, x0, x1)  # Шукаємо дельти для матриці напрямків
        A = A + resOne - resTwo  # Шуаємо власне матрицю напрямків

        x0 = x1
        print_interim_results(x0, calls_of_functions, iteration, restart)
        print("\n")
        iteration += 1

    print(f"\n\n\nМінімум функціі: {x1}")


In [ ]:
dfp([-1.2,0.0], 0.00001, 0.001)

[-1.2, 0.0]
------------------------
Нова точка: [0.94017265 0.88609793]
Функція була обчислена: 64 разів
Ітерація ДФП: 1


Нова точка: [0.9407137  0.88480521]
Функція була обчислена: 89 разів
Ітерація ДФП: 2


Нова точка: [0.98029471 0.95880975]
Функція була обчислена: 144 разів
Ітерація ДФП: 3


Нова точка: [0.99384308 0.98808045]
Функція була обчислена: 198 разів
Ітерація ДФП: 4


Нова точка: [0.99923215 0.9983774 ]
Функція була обчислена: 251 разів
Ітерація ДФП: 5


Нова точка: [1.00000801 1.00001541]
Функція була обчислена: 288 разів
Ітерація ДФП: 6


Нова точка: [1.00000007 1.00000015]
Функція була обчислена: 319 разів
Ітерація ДФП: 7


Кінець програми
Функція була обчислена: 323 разів
ДФП ітерацій проведено: 8
Рестартів у ДФП проведено: 0
Обернена матриця Гессе:
array([[0.49603257, 0.99207017],
       [0.99207017, 1.9891504 ]])
Матриця напрямків А:
array([[0.49798069, 0.99571281],
       [0.99571281, 1.99589643]])



Мінімум функціі: [1.00000007 1.00000015]


## Висновки
У даній курсовій роботі був реалізований метод Девідона-Флетчера-Пауелла на мові Python. Також було проведено низку досліджень, виходячи з яких можна визначити найкращі параметри для ДФП та мінімізувати функцію за найменшу кількість кроків та, власне, оцінки цієї функції. 
Як зазначено вище, було визначено низку параметрів, що якнайкраще підходять для мінімізації функції Розенброка методом ДФП.
* Крок обчислення похідних – 0.00001
* Схема обчислення похідних – центральна
* Метод одновимірного пошуку - ДСК-Пауелла 
* Точність методу одновимірного пошуку – 0.001
* Параметр у алгоритмі Свенна – 0.01
* Точність методу - 0.001

Також хотілося б відмітити, що мінімум функції [1.00000007 1.00000015] знайдений за 323 оцінки функції.

Більш того, матриця напрямків $A$ практично дорівнює $H^{-1}$

Матриця напрямків:
\begin{pmatrix}
0.49798069 &  0.99571281\\
0.99571281 & 1.99589643
\end{pmatrix}

Обернена матриця Гессе:
\begin{pmatrix}
0.49603257 & 0.99207017\\
0.99207017 & 1.9891504
\end{pmatrix}

Для умовної оптимізації був використан метод зовнішніх штрафів. Дослідження показало, що в в незалежності від виду допустимої області, місця знаходження точки функція досягає границі допустимої області, вирівши дві задачі безумовної оптимізації.



## Список використаної літератури
* Методы оптимизации https://math.semestr.ru/optim/optim-manual.php
* НЕЛИНЕЙНОЕ ПРОГРАММИРОВАНИЕ http://www.e-biblio.ru/book/bib/02_estestv_nauki/Matem_Advance/Part9.html
* Химмельблау Д. Прикладное нелинейное программирование / Химмельблау Д. — М. : Мир, 1975. — 535 с.
* Квазиньютоновские методы, или когда вторых производных для Атоса слишком много(Квазиньютоновские методы оптимизации) https://habr.com/ru/post/470638/
* http://nsft.narod.ru/Programming/dfp.html